modules/contracts/apps/challenge/src/handlers/execute.rs

```diff
@@ -0,0 +1,385 @@
+    let final_vote = vote.unwrap_or(true);
+
+    // Construct the vote
+    let vote_entry = Vote {
+        voter: info.sender.to_string(),
+        vote: Some(final_vote),
+        challenge_id: challenge_id.to_owned(),
+    };
+
+    // Load existing votes for the current block height or initialize an empty list if none exist
+    let mut votes_for_block = VOTES
+        .load(deps.storage, challenge_id.to_owned())
+        .unwrap_or_else(|_| Vec::new());
+
+    // Append the new vote
+    votes_for_block.push(vote_entry);
+
+    // Save the updated votes
+    VOTES.save(deps.storage, challenge_id.to_owned(), &votes_for_block)?;
+
+    Ok(Response::new().add_attribute("action", "cast_vote"))
+}
+
+fn count_votes(
+    deps: DepsMut,
+    info: MessageInfo,
+    _env: Env,
+    app: &ChallengeApp,
+    challenge_id: String,
+) -> AppResult {
+    let votes_for_challenge = VOTES
+        .load(deps.storage, challenge_id.clone())
+        .unwrap_or_else(|_| Vec::new());
+
+    let any_false_vote = votes_for_challenge.iter().any(|v| v.vote == Some(false));
+    if any_false_vote {
+        return charge_penalty(deps, info, app, challenge_id);
+    }
+
+    Ok(Response::new().add_attribute("action", "count_votes"))
+}
+
+fn charge_penalty(
+    deps: DepsMut,
+    info: MessageInfo,
+    app: &ChallengeApp,
+    challenge_id: String,
+) -> AppResult {
+    let challenge = CHALLENGE_LIST.load(deps.storage, challenge_id.clone())?;
+    let friends = CHALLENGE_FRIENDS.load(deps.storage, challenge_id.clone())?;
+
+    let admin_address = deps.api.addr_validate(info.sender.as_ref())?;
+    let bank = app.bank(deps.as_ref());
+    let executor = app.executor(deps.as_ref());
+
+    match challenge.collateral {
+        Penalty::FixedAmount { asset } => {
+            let num_friends = friends.len() as u128;
+            if num_friends == 0 {
+                return Err(AppError::Std(StdError::generic_err(
+                    "No friends found for the challenge.",
+                )));
+            }
+
+            // Calculate each friend's share
+            let amount_per_friend = asset.amount.u128() / num_friends;
+            let asset_per_friend = OfferAsset {
+                name: asset.name,
+                amount: Uint128::from(amount_per_friend),
+            };
+
+            // Create a transfer action for each friend
+            let transfer_actions: Result<Vec<_>, _> = friends
+                .into_iter()
+                .map(|friend| bank.transfer(vec![asset_per_friend.clone()], &friend.address))
+                .collect();
+
+            let transfer_msgs = executor.execute(transfer_actions?);
+
+            Ok(Response::new()
+                .add_messages(transfer_msgs)
+                .add_attribute("action", "charge_fixed_amount_penalty"))
+        }
+        Penalty::Daily {
+            asset,
+            split_between_friends: _,
+        } => {
+            // Not sure what the exact implementation should be here.
```

Comment:
I'm not sure either... An optional end-date for the challenge would be nice to have tho.